validation/fpc/fpc_validation_logic.go

```diff
@@ -0,0 +1,340 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package fpc
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/hyperledger-labs/fabric-private-chaincode/ecc/crypto"
+	"github.com/hyperledger-labs/fabric-private-chaincode/internal/utils"
+	"github.com/hyperledger/fabric-protos-go/common"
+	commonerrors "github.com/hyperledger/fabric/common/errors"
+	"github.com/hyperledger/fabric/common/flogging"
+	vs "github.com/hyperledger/fabric/core/handlers/validation/api/state"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/rwsetutil"
+	"github.com/hyperledger/fabric/protoutil"
+)
+
+var logger = flogging.MustGetLogger("fpc-vscc")
+
+// New creates a new instance of the ercc VSCC
+// Typically this will only be invoked once per peer
+func New(sf vs.StateFetcher) *FPCValidator {
+	return &FPCValidator{
+		verifier:     &crypto.ECDSAVerifier{},
+		stateFetcher: sf,
+	}
+}
+
+type FPCValidator struct {
+	verifier     crypto.Verifier
+	stateFetcher vs.StateFetcher
+}
+
+type validationArtifacts struct {
+	chaincodeId string
+	args        [][]byte
+	rwset       []byte
+	response    []byte
+}
+
+// Validate validates the given envelope corresponding to a transaction with an endorsement
+// policy as given in its serialized form
+func (vscc *FPCValidator) Validate(block *common.Block, namespace string, txPosition int, actionPosition int, policy []byte) commonerrors.TxValidationError {
+
+	// get first all information we need to validate enclave endorsenebt
+	va, err := vscc.extractValidationArtifacts(block, txPosition, actionPosition)
+	if err != nil {
+		return policyErr(err)
+	}
+
+	// finally validate proposal and response
+	txverr := vscc.validateFPCEndorsement(va)
+	if txverr != nil {
+		logger.Errorf("VSCC error: validateFPCEndorsement failed, err %s", txverr)
+		return txverr
+	}
+	return nil
+}
+
+func (vscc *FPCValidator) extractValidationArtifacts(
+	block *common.Block,
+	txPosition int,
+	actionPosition int,
+) (*validationArtifacts, error) {
+	// get the envelope...
+	env, err := protoutil.GetEnvelopeFromBlock(block.Data.Data[txPosition])
+	if err != nil {
+		logger.Errorf("VSCC error: GetEnvelope failed, err %s", err)
+		return nil, err
+	}
+
+	// ...and the payload...
+	payl, err := protoutil.UnmarshalPayload(env.Payload)
+	if err != nil {
+		logger.Errorf("VSCC error: GetPayload failed, err %s", err)
+		return nil, err
+	}
+
+	chdr, err := protoutil.UnmarshalChannelHeader(payl.Header.ChannelHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate the payload type
+	if common.HeaderType(chdr.Type) != common.HeaderType_ENDORSER_TRANSACTION {
+		logger.Errorf("Only Endorser Transactions are supported, provided type %d", chdr.Type)
+		err = fmt.Errorf("Only Endorser Transactions are supported, provided type %d", chdr.Type)
+		return nil, err
+	}
```

Comment:
This is something we haven't done before. From `vendor/github.com/hyperledger/fabric-protos-go/common/common.pb.go` no other type seems relevant (and that's why we probably always received only endorser transactions) but definitely robust programming to explicitly check ...

Reply:
this check we inherited from also from the dafaultValidator https://github.com/hyperledger/fabric/blob/master/core/handlers/validation/builtin/v20/validation_logic.go#L109